pinc/ProofProject.inc

```diff
@@ -0,0 +1,217 @@
+                    $this->round->id
+                );
+                throw new DailyLimitExceededException($sentence);
+            } else {
+                $this->save_as_done();
+                if ($pre_save_dpl_count + 1 >= $this->round->daily_page_limit) {
+                    $sentence = sprintf(
+                        _("Your page has been saved as 'Done'. However, you have now reached the daily page limit for %s."),
+                        $this->round->id
+                    );
+                    throw new DailyLimitReachedException($sentence);
+                }
+            }
+        } else {
+            $this->save_as_done();
+        }
+    }
+
+    private function save_as_done(): void
+    {
+        global $pguser;
+
+        $this->page_state = Page_saveAsDone($this->projectid, $this->page_name, $this->round, $pguser, $this->page_text);
+        // add to user page count
+        page_tallies_add($this->round->id, $pguser, +1);
+    }
+
+    public function save_as_in_progress(string $page_text): array
+    {
+        global $pguser;
+        $this->page_text = $this->get_valid_text($page_text);
+        $this->saveAsInProgress($page_text, $pguser);
+        return $this->get_page_text_array() + $this->get_page_state_array();
+    }
+
+    public function save_and_revert(string $page_text): array
+    {
+        global $pguser;
+        $this->page_text = $this->get_valid_text($page_text);
+        $this->saveAsInProgress($page_text, $pguser);
+        $this->revertToOriginal();
+        return $this->get_page_text_array() + $this->get_page_state_array();
+    }
+
+    public function resume_page(): array
+    {
+        global $pguser;
+
+        // page could be in out, temp or saved state
+        if ($this->page_state == $this->round->page_save_state) {
+            // Page comes from DONE.
+            // no need to update text, just record state change
+            $this->page_state = Page_reopen($this->projectid, $this->imagefile, $this->round, $pguser);
+            // When it was saved, the user's page-count was incremented.
+            // Now they are 'unsaving' it, so decrement their page-count.
+            // They'll get it back if/when they save-as-done again.
+            page_tallies_add($this->round->id, $pguser, -1);
+        } else {
+            // Page comes from IN PROGRESS or OUT.
+            // Resuming such a page has no actual effect on the page,
+            _Page_require($this->projectid, $this->page_name, [$this->round->page_out_state, $this->round->page_temp_state], $this->round->user_column_name, $pguser, 'resume');
+        }
+        return $this->render_page_data();
+    }
+
+    public function render_page_text(): array
+    {
+        // validate user and page? page should be in out or temp state only
+        return $this->get_page_text_array();
+    }
+
+    private function render_page_data(): array
```

Comment:
This duplicates some of the function of PPage.inc. Why not make a function in LPage to get the data which we can also use in the api and use it in PPage to echo the data?

Reply:
Or make a separate PR to do that.